### PR TITLE
Can check isValidMove except for special moves

### PR DESCRIPTION
**Pull request by nk18chi**
Can check if the piece you try to move is owned by you
Can check if the destination is occupied by your piece
Can check if a piece exists on the way between current position and destination 
Can check if a piece can move as basic move

### CODE REVIEW

File: src/chess/SquareManageable.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package chess;
+
+import chess.piece.Piece;
+
+public interface SquareManageable {
+  public Piece get(int row, int col);
+  public void update(int fromRow, int fromCol, int toRow, int toCol);
+  public void remove(int row, int col);
+  public Piece[] getAttackedList(int row, int col);
+}
```
Comment (usk83) [inline src/chess/SquareManageable.java]: This interface shouldn't be in `chess` package but `chess.piece`

File: src/chess/SquareManageable.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package chess;
+
+import chess.piece.Piece;
+
+public interface SquareManageable {
+  public Piece get(int row, int col);
+  public void update(int fromRow, int fromCol, int toRow, int toCol);
+  public void remove(int row, int col);
+  public Piece[] getAttackedList(int row, int col);
```
Comment (usk83) [inline src/chess/SquareManageable.java]: Maybe `getAttackablePieces` mekes more sense

File: src/chess/Board.java [hunk 3]
```diff
@@ -35,8 +35,102 @@ public boolean update(Position pos, Position newPos) {
...
+
+  @Override
+  public void remove(int row, int col) {
+
+  }
+
+  @Override
+  public Piece[] getAttackedList(int row, int col) {
+    return new Piece[0];
+  }
```
Comment (usk83) [inline src/chess/Board.java]: You should put `TODO: xxx` comments on incompleted codes

File: src/chess/Board.java [hunk 4]
```diff
@@ -35,8 +35,102 @@ public boolean update(Position pos, Position newPos) {
     return true;
   }
 
+  public boolean isPiece(Position pos) {
```
Comment (usk83) [inline src/chess/Board.java]: private?
public or private is very important because if a method is public, others think its method is for other classes

Comment (usk83) [reply src/chess/Board.java]: Basically, when you're not sure it should be private or public, you should put private first
and when you really need that method to call from the other class, then you can modify it to public

File: src/chess/Board.java [hunk 5]
```diff
@@ -35,8 +35,102 @@ public boolean update(Position pos, Position newPos) {
...
+    return metrix[pos.getCol()][pos.getRow()] != null;
+  }
+
+  public boolean isOwnPiece(Position pos, Color c) {
+    if (!this.isPiece(pos)) {
+      return false;
+    }
+
+    Piece that = metrix[pos.getCol()][pos.getRow()];
```
Comment (usk83) [inline src/chess/Board.java]: you have almost same code on line 39
So most of the case you can make it better
This time, I guess you don't need to implement `isPiece` but `getPiece` instead

File: src/chess/Board.java [hunk 6]
```diff
@@ -35,8 +35,102 @@ public boolean update(Position pos, Position newPos) {
...
+    int x = from.getRow() - to.getRow();
+    int y = from.getCol() - to.getCol();
+    int x_abs = Math.abs(x);
+    int y_abs = Math.abs(y);
+
+    // if less than two spaces between the piece and destination
+    if (Math.max(x_abs, y_abs) < 2) return true;
+
+    if ((x_abs == y_abs) || (x_abs == 0) || (y_abs == 0)) {
+
```
Comment (usk83) [inline src/chess/Board.java]: remove this blank line

File: src/chess/Board.java [hunk 7]
```diff
@@ -35,8 +35,102 @@ public boolean update(Position pos, Position newPos) {
...
+    }
+
+    return true;
+  }
+
+  public boolean ableBasicMove(Position from, Position to, Color c) {
+    int x = from.getRow() - to.getRow();
+    int y = from.getCol() - to.getCol();
+    Piece p = metrix[from.getCol()][from.getRow()];
+    return p.isValidMove(x, y, isEnemyPiece(to, c));
```
Comment (usk83) [inline src/chess/Board.java:111]: Maybe you also need to consider which way does the color facing (you know, piece have a direction)
But this is enough so far:+1:

File: src/chess/Game.java [hunk 8]
```diff
@@ -62,14 +63,38 @@ public void play() {
...
+          // Check Halfway
+          if (!board.isNotPiecesOnHalfway(fromPosition, toPosition)) {
+            System.out.println("Can not move the selected piece because other piece is on halfway.");
+            break;
+          }
+
+          // Check Basic move of a Piece
+          if(!board.ableBasicMove(fromPosition, toPosition, TUNES[turnIndex])) {
+            System.out.println("The piece you selected doesn't allow to move to the destination");
+            break;
```
Comment (usk83) [inline src/chess/Game.java]: This is validation for updating, so maybe should be done within `board.update` method
but you can keep left this code for now:+1:

File: src/chess/SquareManageable.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package chess;
+
+import chess.piece.Piece;
+
+public interface SquareManageable {
+  public Piece get(int row, int col);
+  public void update(int fromRow, int fromCol, int toRow, int toCol);
+  public void remove(int row, int col);
```
Comment (usk83) [inline src/chess/SquareManageable.java]: I think it better to return the removed piece if exist for a future situation

File: src/chess/piece/Piece.java [hunk 10]
```diff
@@ -18,17 +18,19 @@ public Piece(Color color, Map<Color, Character> symbols) {
 
     this.color = color;
     symbol = s;
+    this.lastMovedTurn = 0;
```
Comment (usk83) [inline src/chess/piece/Piece.java]: `this` is unnecessary

File: src/chess/piece/Piece.java [hunk 11]
```diff
@@ -18,17 +18,19 @@ public Piece(Color color, Map<Color, Character> symbols) {
...
-        && newPosition.getCol() > 0
-        && newPosition.getRow() < 8
-        && newPosition.getCol() < 8) {
-      return true;
-    } else {
-      return false;
-    }
+  public Color getColor() {
+    return color;
+  }
```
Comment (usk83) [inline src/chess/piece/Piece.java]: `protected` field is accessible from child classes, so you don't need this

File: src/chess/piece/Piece.java [hunk 12]
```diff
@@ -18,17 +18,19 @@ public Piece(Color color, Map<Color, Character> symbols) {
...
+
+  public void setLastMovedTurn(int lastMovedTurn) {
+    this.lastMovedTurn = lastMovedTurn;
+  }
```
Comment (usk83) [inline src/chess/piece/Piece.java]: same with `getColor`

File: src/chess/piece/Piece.java [hunk 13]
```diff
@@ -18,17 +18,19 @@ public Piece(Color color, Map<Color, Character> symbols) {
...
+
+  public boolean isValidMove(int x, int y, boolean isEnemyExisted) {
+    return true;
```
Comment (usk83) [inline src/chess/piece/Piece.java]: This doesn't make sense!
You can just define this as an abstract method without any content
So you don't need to call this as `super.isvalidMove()` from children

File: src/chess/piece/Piece.java [hunk 14]
```diff
@@ -40,4 +42,5 @@ public void printClassName() {
   public String toString() {
     return String.valueOf(symbol);
   }
+
```
Comment (usk83) [inline src/chess/piece/Piece.java]: unnecessary change

File: src/chess/piece/Pawn.java [hunk 15]
```diff
@@ -12,16 +12,37 @@
         }
       };
 
+
```
Comment (usk83) [inline src/chess/piece/Pawn.java]: blank

File: src/chess/piece/Pawn.java [hunk 16]
```diff
@@ -12,16 +12,37 @@
...
+    if (y < 0 || 2 < y) return false;
+
+    // Diagonally forward
+    if (isEnemyExisted && y == 1 && Math.abs(x) == 1) return true;
+
+    // forward two
+    if (!isEnemyExisted && lastMovedTurn == 0 && x == 0 && y == 2) return true;
+
+    // forward one
+    if (!isEnemyExisted && x == 0 && y == 1) return true;
```
Comment (usk83) [inline src/chess/piece/Pawn.java]: This calculation should be done as `Board` responsibility, but acceptable for now:+1:

**Review by usk83 — COMMENTED** [review]
Big thanks for your lots of great work:+1:

File: src/chess/piece/Piece.java (hunk 11 above)
Comment (nk18chi) [reply src/chess/piece/Piece.java]: This might be necessary when I talked to Yusuke. So, I don't change anything

File: src/chess/piece/Piece.java (hunk 12 above)
Comment (nk18chi) [reply src/chess/piece/Piece.java]: This might be necessary when I talked to Yusuke. So, I don't change anything

File: src/chess/Board.java (hunk 7 above)
Comment (nk18chi) [reply src/chess/Board.java:111]: I didn't improve it. Please polish it 🙇

File: src/chess/Game.java (hunk 8 above)
Comment (nk18chi) [reply src/chess/Game.java]: I didn't improve it. Please polish it, too 🙇

File: src/chess/piece/Pawn.java (hunk 16 above)
Comment (nk18chi) [reply src/chess/piece/Pawn.java]: I didn't improve it. Please polish it 🙇

File: src/chess/piece/Piece.java (hunk 11 above)
Comment (usk83) [reply src/chess/piece/Piece.java]: You're right! My fault

File: src/chess/piece/Piece.java (hunk 12 above)
Comment (usk83) [reply src/chess/piece/Piece.java]: My fault again:bow:

**Review by usk83 — APPROVED** [review]
now it's ready to merge!
